### ISSUE TITLE
Is it acceptable to use HTML for the serialization of some infoset items, or should it all be in separate (JSON) file?

### ISSUE THREAD

**Issue by iherman** [w3c/wpub#193]
This discussion has permeated many of the various issues (e.g., lately, #159, #181, or #186). It would help to get this design principle settled once and for all. In practice, the issue is whether the "entry (HTML) page" could be used as containing the infoset items, or not.

Note that the answer may not be clear-cut, and may depend on the nature of the infoset items. Indeed, it is different if:

1. the item is, loosely speaking, some sort of a metadata, i.e., expressible via an HTML `<meta>` or `<link>` element (e.g., creation modification date or links to an ONIX file)
2. the item is a slightly more complex structure that cannot be expressed fully in the HTML header (e.g., language and base direction)
3. the item is an item that would be naturally expressed in HTML, and is often indeed done that way (e.g., Table of Content)

Another aspect that influences this decision is whether the WP consists of a single HTML file (which is also the entry page), with adjunct files like CSS or images. This is the typical case for, e.g., a scholarly journal article.

**Comment by llemeurfr** [comment]
Note that editing such infoset **by hand** would be equally difficult in the html and json cases. Meaning that whatever the choice is btw a highly specific web page and a json structure, an authoring tool seems mandatory. Which leads to an additional Con.

* **Con**: creating an authoring tool capable of generating an html page both highly semantic (metadata, reading order, tables of illustrations, list of additional resources etc.) and highly flexible regarding its layout would be really difficult. In comparison, creating a tool capable of generating a json structure + letting authors use html tools to create an entry page is a breeze.

**Comment by HadrienGardeur** [comment]
Just a few quick notes first:

- we've already agreed on using JSON as our manifest format
- which mean that we're discussing if **some** infoset items can be contained in HTML resources from the publication, not **all** infoset items
- the list provided by @iherman assumes that additional infoset items would be expressed on the "entry page" (I hate that term), but it's very doubtful that all navigation (not just the TOC but also page lists, landmarks and various other lists) for example would be listed on a single page or on the entry page

I'd also like to list an additional con: may require additional network requests that could block the processing of the WP.

If I discover a publication through one of its chapter, this means that:

- I'll need to fetch the manifest first
- then I'll need to fetch one or more HTML resources before I have all the info that I need to properly render the publication and/or provide the relevant affordances

Since I'll only be able to discover these additional HTML resources through the manifest, this means that these fetch requests (plus all the processing related to HTML) will have to be done sequentially and not in parallel.

The majority of the pros listed by @iherman could also be challenged IMO because they're mixing up two different issues:

- how we serialize this information
- vs how we access it

In the case of a single-HTML document, I don't think that using  `<meta>` + `<link>`+ potentially RDFa is in any way better than just embedding JSON-LD in the HTML document.

There are less semantic issues with JSON-LD (the metadata is not necessarily about the document that contains them) and I would argue that it's easier to author JSON-LD than RDFa.

To go back to the list of pros, we could also say that JSON-LD embedded in HTML is:

- a natural way of expressing metadata for single page publications (cf [AMP](https://www.ampproject.org/docs/fundamentals/discovery#use-schema.org-for-most-search-engines))
- provides a simple structure as well

I'm not really buying the redundancy arguments (we're not expressing the same information) or the more "natural bridge" one (browsers ignore the vast majority of metadata and links that we would end up using in HTML).

I'd like to hear @BCWalters opinion on this as well, now that we have a major browser actively participating in this WG, I think there's a lot of value to what they have to say about this.

**Comment by RachelComerford** [comment]
There is a business consideration that weighs into the HTML vs JSON question because it is easier and cheaper for me to find HTML coding resources than JSON coding resources and my team is less likely to follow a standard that is (even more) expensive to maintain. To confirm this, I reached out to our most commonly used vendors - all replied that they would need time to staff up and train JSON developers but that they had plenty HTML developers on staff.

**Comment by iherman** [comment]
Thanks @RachelComerford, this is a very important, non-technical point...

**Comment by BigBlueHat** [comment]
@iherman
> the item is, loosely speaking, some sort of a metadata, i.e., expressible via an HTML <meta> or <link> element (e.g., creation modification date or links to an ONIX file)

I'd not limit it to *just* `<meta>` and `<link>`. The growth and widespread usage of data-in-HTML formats (RDFa, Mircodata, JSON-LD) show that developers and web publishers *do* know how to put metadata in their publications and apps, and are *already* incentivized to do so because search engines. Why not follow suite rather than creating a different, currently unknown, out-of-band location to look for metadata?

**Comment by deborahgu** [comment]
I'd like to make another non-technical point: we should not be creating a complex creation systems for publishers. Descriptive metadata, including navigation items, should go in as few files, and as few formats, as is technically possible.  

As Ivan said:

> Pro: mainly in the case of single HTML file based WP this is a natural way of expressing the information.

If we tell publishers "in order to create a WP, you need to put this infoset data over *here* in HTML, and this infoset data over *there* in JSON," we're raising the barrier to entry for anyone who doesn't have a WP-aware authoring tool.  

IMO, much better to choose an imperfect design which publishers will actually be able to use than the most perfectest beautifullest awesomest architecture which is a pain for creators.

(I have no horse in the race of actual location and format, and personally I'd be happiest if all the players in this conversation came to a place where they realize that no solution is perfect and all the people disagreeing have valid points. Unfortunately a classic compromise is the worst possible solution, because we really just need to pick one. There is literally no solution on offer without cons; we still have to choose one and move on to the rest of the work.)

**Comment by iherman** [comment]
@BigBlueHat

> @iherman
>> the item is, loosely speaking, some sort of a metadata, i.e., expressible via an HTML <meta> or <link> element (e.g., creation modification date or links to an ONIX file)
>
> I'd not limit it to *just* `<meta>` and `<link>`. The growth and widespread usage of data-in-HTML formats (RDFa, Microdata, JSON-LD) show that developers and web publishers *do* know how to put metadata in their publications and apps, and are *already* incentivized to do so because search engines. Why not follow suit rather than creating a different, currently unknown, out-of-band location to look for metadata?

I know there can be more data than just `<meta>` or `<link>`. But I believe the characterization of a "different, currently unknown, out-of-band location to look for metadata" is a bit harsh. Putting metadata into a separate file, and link to it, is not a new approach, see (beyond the WAM) the work on [Payment Method Manifest](https://www.w3.org/TR/payment-method-manifest/), and was also the routine approach to get to metadata before the creation of RDFa, with the metadata stored in different formats, let that be Turtle or (God forbid!) RDF/XML. (This was, e.g., the way to refer to CC metadata from an HTML page.)

**Comment by iherman** [comment]
@BigBlueHat,

> > Con: per the HTML standard, the <meta> element's role is to express "document-level metadata" (see html5; emphasis is mine). Using it for expressing metadata > for other entities (ie, the WP) is semantically not clean. (In the case of a single HTML file based WP one could argue that the document and the HTML file is > the same, which would make it all right.)
> 
> Depending on how this is modeled and "gone about" it maybe that the "binding" document is imperceptible from the publication itself.
> 

Yes, I agree; this is the case of a single-document PW; this is one of the "Pro" arguments.

> Or, alternatively:
> * publication address might be `http://example.com/moby-dick/`
> * binding document (currently "entry point") is returned upon that request (i.e. `index.html` per most server defaults), but has it's *own* URL `http://example.com/moby-dick/index.html` and consequently could have it's *own* metadata (in `<meta>` or wherever).

Sorry, but I do not agree. The quoted HTML specification does not refer to a URL, it refers to the document itself, whichever path was used to get there. I believe the HTML standard is pretty clear about it. If we use the HTML headers, we should simply accept that we are willfully overstepping the bounds that the HTML standard defines (but I am not sure the rest of the community would accept it, we may face major objections).

> 
> Regardless, this is easily avoidable...so no an implicit "con."

I think we have to agree that we disagree on that point.

> 
> > Con: the metadata may be used for various purposes handling the WP instance itself, eg, indexing, bookshelves, etc. Parsing an HTML file to extract the  information, though would use a standard toolset, requires a significant effort for the User Agent: parsing the HTML, building the DOM, the CSS DOM, the  Accessibility DOM, etc, before giving access to the <meta> element. Compared to that, parsing a JSON file into Javascript structures is a breeze.
> 
> There's no *requirement* that a DOM, CSSOM, Accessibility OM, etc. be setup or available when extracting metadata from HTML files. It's possible to get it directly out of the markup *without* those things.

This is theoretically correct, but I do not think it is practically true. Any implementation will use one of the many, possibly "built-in" HTML parsers, and all those parsers build up the DOM. I do not think we can expect an implementation to have a _different_ parser that would just look at the syntax or do some other tricks.

> 
> Additionally, when "browsed to" the browser will provide all those things, and could potentially make that data more easily extract-able by the developer (or within the UI of the browser).

I am not sure I understand what you mean. Yes, of course, if the UA begins to render, display, etc, the WP, then those data are already there, because they are in the DOM. The "Con" is for the cases when, say, the Reading System or the browser builds up, say, bookshelf, for which a number of Infoset items are necessary.

That being said, if we go along with the idea of finding the manifest file via a `<link>` element, then the same problem applies. So this may be one of the 'con'-s that we have to live with whatever we do, and we can consider it neutral in our discussions:-)

> > Con: having both a manifest file and some data in some HTML resources complicates implementations that should follow a more complex path to get hold of the > infoset item. (Note, however, that this argument has less weight than ease of authoring; there are more authors than implementers...)
> 
> Couldn't agree more...but that's not a "con" of an HTML-driven approach to these problems.

True... except that it remains to be proven that _all_ infoset items can be expressed easily and in a user-friendly manner via the current HTML element set. 

To take an example: we did say that the language tag in a content file (ie, an HTML file) is not the same as the language tag for the publication as a whole. In other words, the regular `@lang` attribute in the HTML file _cannot_ be used as an encoding of the relevant infoset items: it has to be put somewhere else. We will have to define our own definitions for that in some way or other, which will not look very natural in HTML (and hence not very user friendly). We may have similar issues with, say, the title of the WP...

> 
> If, for instance, all the primary resources are referenced from an HTML-based "binding document" (perhaps through something like a latent-loading `<iframe>` or a `<nav role="doc-toc">` *like* thing), then the request and processing needs are already defined and taken care of by the browser and the HTTP ecosystem specs (CORS, CSP, etc). However, if they're in the JSON (as noted in #104), there's an unknown relationship with the things stated there and the rest of the request/response processing constraints, browsing contexts, etc (again; hence #104). So...that's ultimately a "vote" for primary resources to be expressed from within the HTML.

To be honest, you lost me here; more exactly, I do not see the problem. If we say (as we seem to converge to in #104) that we simply take the browsing context as given, I just do not see the issue accessing the separate JSON file in this browsing context. That information is accessed from the entry point (in its own browsing context, as we seem to converge to in #104), then all the rest is clear: _that_ is the context we are operating in. Let alone the fact that many elements in the infoset (title, authors, etc) are unaffected by the browsing context.

> Each of the current infoset items are expressible from within an HTML document (see my last comment for a handful of options), and what's needed next is to know how to *enhance* their expressions as available now such that they are *more* useful.

See my comment above. I am absolutely not sure it is as simple, more exactly that the resulting definitions would be clearer and simpler than doing it in JSON.

Note that the experience in RDFa is not really good (alas!), meaning the relying on RDFa may not be that helpful. (Authoring RDFa can be a major challenge, and is very opaque for non RDF-savy persons (and is sometimes difficult even for people like me, I frequently have to run RDFa+HTML through my own distiller to see what the generated RDF is). Microdata is, maybe, even worse, because there are features that cannot even be expressed in microdata...)

**Comment by iherman** [comment]
Trying to move forward: would the usage of a `<script>` element alleviate the problems? (See also #122). Here is what this would mean:

- we define the manifest in JSON (LD or not); that would define many? most? all? of our Infoset elements (not "all" would meant that some entries are put into an HTML content of the entry point somewhere)
- this JSON content _may_ be a separate file referred to from the entry point and linked to via a special `<link>` element as defined in the draft
- this JSON content _may_ also be part of the entry point file, encapsulated in a `<script>` element. 

What this means is that there is not _necessarily_ a separate file to be authored; all is in the same file; would that alleviate your issues, @deborahgu and @RachelComerford ? It would not necessarily help with the issues of @llemeurfr because today's authoring tools rarely help for the authoring of embedded data. On the other hand, the semantics of the `<script>` element's content is under our control, ie, we would not violate the HTML spec.

The experience shows that authoring JSON for metadata-like information is simpler than doing it in, say, RDFa, so we would gain that.

Note also [DanBri's comment](https://github.com/w3c/wpub/issues/122#issuecomment-359786173): Schema.org also uses this JSON(-LD) wrapper to extract information. 

(An even more radical proposal would be to use the embedded `<script>` element _only_. I am not sure I would go that far.)

**Comment by danielweck** [comment]
Ivan, +1 to the JSON-in-script / JSON-as-file approach (although I suspect reading system developers would prefer a directly-accessible standalone JSON, as this saves parsing an HTML document and performing an additional fetch request).

**Comment by dauwhe** [comment]
> Sorry, but I do not agree. The quoted HTML specification does not refer to a URL, it refers to the document itself, whichever path was used to get there. I believe the HTML standard is pretty clear about it. If we use the HTML headers, we should simply accept that we are willfully overstepping the bounds that the HTML standard defines (but I am not sure the rest of the community would accept it, we may face major objections).

Consider the following document returned from `www.example.com/book/`

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <title>Moby-Dick</title>
  <meta name="author" content="Herman Melville">
</head>
<body>
  <nav>
    <ol>
      <li><a href="c1.html">One</a></li>
      <li><a href="c2.html">Two</a></li>
    </ol>
  </nav>
  <iframe id="c1" name="c1" src="c1.html"></iframe>
  <iframe id="c2" name="c2" src="c2.html"></iframe>
</body>
</html>
```

If `c1.html` does not have a `meta name="author"` element, who is the author of `c1.html`? The content of `c1.html` is literally a node in the document object of the original URL. Would the answer be different if `c1.html` was included via `object`, html imports, or a custom element?

**Comment by mattgarrish** [comment]
>> There's no requirement that a DOM, CSSOM, Accessibility OM, etc. be setup or available when extracting metadata from HTML files. It's possible to get it directly out of the markup without those things.
> 
> This is theoretically correct, but I do not think it is practically true.

How are these steps avoided? Is the idea that user agents will go through the process of obtaining and processing the manifest before the user makes any decision about whether they even want to initiate the reading experience, and stop rendering the document until a decision is made?

In other words, does an external file really save anything in processing time, except perhaps in the (rare?) situation where a user says to always initiate publications and the link is available in an HTTP header?

**Comment by llemeurfr** [comment]
@iherman about 

>Note that the experience in RDFa is not really good (alas!), meaning the relying on RDFa may not be that helpful. (Authoring RDFa can be a major challenge, and is very opaque for non RDF-savy persons (and is sometimes difficult even for people like me, I frequently have to run RDFa+HTML through my own distiller to see what the generated RDF is). Microdata is, maybe, even worse, because there are features that cannot even be expressed in microdata...)

I totally agree with that statements. At allocine.com, we embedded RDFa, then microdata (preferred), in our film / star etc. pages. But it was the work of the technical team, in page templates: certainly not the work of the editorial team. And I'm pretty sure that this is how 99.9% of websites containing RDFa or microdata are constructed.

**Comment by iherman** [comment]
@dauwhe (referring to https://github.com/w3c/wpub/issues/193#issuecomment-388031101) great questions...

I am not sure, and I do not think the HTML spec clearly says anything about this case. However, looking at the [HTML spec](https://www.w3.org/TR/html5/semantics-embedded-content.html#the-iframe-element), a document within an iframe has its own Document element (and own context), so my gut feeling is that, in your example, the author of the iframe-d content would be unknown. It is probably the same with `object`. The import case is even less clear, the [current draft](https://www.w3.org/TR/html-imports/) does not really say anything about Document elements or contexts (is that work still alive, b.t.w.?).

**Comment by RachelComerford** [comment]
@iherman... to be honest, I don't understand the solution?

_Trying to move forward: would the usage of a <script> element alleviate the problems? (See also #122). Here is what this would mean:

- we define the manifest in JSON (LD or not); that would define many? most? all? of our Infoset elements (not "all" would meant that some entries are put into an HTML content of the entry point somewhere)
- this JSON content may be a separate file referred to from the entry point and linked to via a special <link> element as defined in the draft
- this JSON content may also be part of the entry point file, encapsulated in a <script> element.

What this means is that there is not necessarily a separate file to be authored; all is in the same file; would that alleviate your issues, @deborahgu and @RachelComerford ?_

**Comment by iherman** [comment]
Sorry to be terse, @RachelComerford. My bad.

The choices discussed so far were:

1. the infoset items are in a separate JSON file, which is linked from the main HTML entry point of the WP using a `<link>` element; or
2. use the HTML entry point to the WP to also "encode", in some way or other, all the infoset entries using, mostly, `<meta>` and `<link>` elements, as well as, possibly, some HTML elements (I presume salt-'n-peppered with RDFa or microdata terms')

Both you and @deborahgu commented that to edit two separate files would be a load on your developers; as @deborahgu said "Descriptive metadata [...] should go in as few files, and as few formats".

I do have some significant problems with the 2nd approach. In essence, I believe, it is not really possible to avoid some extra "formats" (where by format I also mean microdata and/or RDFa, or complex set of attributes on HTML elements, etc.) and, among those, I also believe JSON is still the simplest one. But at least the problem of handling several files can be alleviated. Indeed, the HTML standard allows to use the following in the header of an HTML content:

```
<script type="application/json">
    "something"      : "whatever",
    "somethingselse" : "whateverelse"
</script>
```

I.e., we can use this element to encode the infoset items in JSON, while _staying within the (entry point) HTML file_.

This is how webmasters mark up their files for schema.org, b.t.w., if they choose JSON-LD for their data. (See a [random example](http://schema.org/address): go to the bottom of the page to choose the right tab for some examples.) In other words, we would be in good company:-) and, in fact, the metadata part of our infoset may automatically be used, when on the Web, by schema.org (at least for schema.org terms) which is a bonus...

I hope this is clearer...

**Comment by HadrienGardeur** [comment]
Handling multiple files is only a "problem" for single resource publications.

With publications spread across multiple resources, going through an HTML resource to extract JSON-LD in a script element makes things more complicated than it should be.
I would recommend limiting this embedded case strictly to publications that have no explicit reading order (=single resource publications).

That said, I fully agree that for expressing metadata on an HTML resource, JSON-LD + schema.org is the preferred method on the Web today.
It's much easier to author than RDFa and doesn't have the limitations associated with the `<meta>` element.
If we go down that road, we should be careful how we author such metadata if we want to maximize compatibility with SEO bots. Not all of them are truly JSON-LD/RDF aware and they tend to have limited support for context documents.

**Comment by BigBlueHat** [comment]
> I believe the characterization of a "different, currently unknown, out-of-band location to look for metadata" is a bit harsh.

There was no intention to be "harsh."

I meant the technical term "[out-of-band](https://en.wikipedia.org/wiki/Out-of-band)":
> Out-of-band is activity outside a defined telecommunications frequency band, or, metaphorically, outside some other kind of activity.

And "[out-of-band data](https://en.wikipedia.org/wiki/Out-of-band_data)":
> out-of-band data is the data transferred through a stream that is independent from the main in-band data stream.

"different" and "currently unknown" were meant to be positioned in contrast with where Search Engines (the primary incentivize driving web publication metadata) get their information.

I apologize that the sentence was so easily misread as to contain a "harsh" tone. I'll work to link to any technical terms that may have taken on a different cultural meaning.

**Comment by BigBlueHat** [comment]
@iherman it was never my intention to have this HTML approach be limited in anyway to *just* what is expressible directly on elements and attributes (i.e. RDFa and Microdata).

Using JSON(-LD) (or any other format) in a `<script>` tag is called "data blocks" in the HTML5 spec, and has been around since "forever"--so I'd never expected that to be removed from consideration.

Apologies for tot being clearer about these specifics.

Building up from HTML means we have *everything* HTML provides at our disposal...including JSON. 😄

**Comment by BigBlueHat** [comment]
Additionally, the infoset includes both metadata pieces (title, reading progression, etc) and request/response/hypermedia related things (primary resources, etc).

The point of using HTML as a foundation is that it already has a known (and carefully crafted) set of specifications for handling the inclusion, processing, and contextualizing of Web resources. We don't have that (to my knowledge) with any other format (because SVG doesn't have iframes :wink:).

The metadata concerns and the "binding"/rendering/presenting concerns are different domains of use, experience, security, etc. We should model them accordingly.

**Comment by BigBlueHat** [comment]
> How are these [infoset extraction] steps avoided?

The underpinning premise is that "web publications" (lowercase on purpose) exist today (ex: http://guide.couchdb.org/ http://hpbn.co/ http://resilientwebdesign.com/ )--you can load them up and read them now.

However, their metadata (mostly RDFa because [ogp.me](http://ogp.me/)) and "binding" (which is mostly a next/prev experience) is provided in idiosyncratic ways and either expressed "inline" (next/prev links in each resource) or built via JS (as is the cause with the CouchDB Guide).

Consequently, the "infoset" items are currently expressed with some overlap in consistency in only a few places (mostly ToC and some ogp.me metadata), so there's not much the browser can do to provide *additional* or *enhanced* reading experiences for the publication as a whole (i.e. no publication-wide search unless via a service, linear progression experience is identical to clicking any other link in the book, etc).

> Is the idea that user agents will go through the process of obtaining and processing the manifest before the user makes any decision about whether they even want to initiate the reading experience, and stop rendering the document until a decision is made?

The point is that the experience of the "web publication" (lowercase on purpose again) isn't blocked by anything. Obtaining and processing any additional, consistently expressed data or affordances *could* (in a future browser or via a polyfill in the meantime) *enhance* the reading experience by adding things like publication-wide search, linear progression, etc.

> In other words, does an external file really save anything in processing time, except perhaps in the (rare?) situation where a user says to always initiate publications and the link is available in an HTTP header?

By conceiving of "web publications" (case sensitive one more time :grin:) as *extant* and *enhance-able*, we can lay a foundation to build up from.

In which case, a user choice or action to "initiate [a] publication" would look like (auto)triggering new experiences (search, linear progression, etc) either directly within the browsers UI or perhaps via some dedicated reading UI "space" (or like one of the many things we've not currently imagined 😃). The provision of those *enhancements* though, should not block the Web experience already available to readers of "web publications." The reader's lives should only get better from our work. 😸

**Comment by deborahgu** [comment]
I think my concern, and I expect @RachelComerford's as well, are partly encapsulated by something @iherman said (emphasis mine):

> Both you and @deborahgu commented that to edit two separate files would be a load on your **_developers_**; as @deborahgu said "Descriptive metadata [...] should go in as few files, and as few formats".

It's not the case that existing digital publications are currently primarily put together by developers. In the case of epubs, many are put together by production and editorial staff who have learned enough technical skills to compose the necessary HTML, OPF, etc. files. I'm not saying that these staff can't learn further technical skills; many of us know (or are!) those originally non-technical staff who became technology experts on the job. But many it's not even remotely the case that every organization creating digital publishing has developers responsible for the authoring process.

To phrase what I said another way, if we think the technically correct solution for WPs means that the overhead for creating WPs will either mean using an authoring tool or having developers on staff, we need to acknowledge that will be a possibly insurmountable adoption roadblock for many small organizations.

This is why I think @HadrienGardeur's comment, 

>Handling multiple files is only a "problem" for single resource publications.

Is only true if our focus is entirely user agent focused. I worry about _production_.

I'd be happiest if the major proponents of each method created a pseudocode mockup their representation of the infoset from, say, https://github.com/IDPF/epub3-samples/tree/master/31/moby-dick-mo and linked them here, so we could see what they'd look like from an authoring standpoint.  (I believe some of you have already created such mockups. 😄)

**Comment by BigBlueHat** [comment]
@deborahgu thank you for the context setting! It's appreciated.

Here's a more recent attempt at an HTML-first/only Moby-Dick--this one based *entirely* on the files in https://github.com/IDPF/epub3-samples/tree/master/31/moby-dick-mo/EPUB

**Warning: Chrome only!** (for now)
https://wileylabs.github.io/no-can-transclude/moby-dick-from-epub-samples/
 - it's the exact files from [the `/EPUB/` folder in the Moby-Dick sample](https://github.com/IDPF/epub3-samples/tree/master/31/moby-dick-mo/EPUB)
 - the *only* change to the files is renaming `toc.xhtml` to `index.html` and adding one script tag to provide the polyfill-like reader overlay: https://github.com/WileyLabs/no-can-transclude/commit/5c06b39a59206afdc7cf9347c95c36fb40e0f4d5#diff-0
 - uses `<dialog>` to create an overlay reading space (hence the need for Chrome...)

There's much more polyfilling to do to address other concerns (direct resource identification in context, etc), but it's one of several promising attempts. I'll share more later in other issues/threads. This one focuses nicely on the "coming from EPUB" story, hopefully.

**Comment by deborahgu** [comment]
@BigBlueHat, we all agree that web publications exist today, but I don't think we can pick a set of publications that do things in a certain way and say those show all needs can be easily met. It's true that http://guide.couchdb.org/ is a web publication, but so is [this Buzzfeed article](https://www.buzzfeed.com/alivelez/the-passion-of-rogelio-de-la-vega), each of these [fan fiction stories about the publishing industry](https://archiveofourown.org/works?utf8=%E2%9C%93&commit=Sort+and+Filter&work_search%5Bsort_column%5D=revised_at&work_search%5Brating_ids%5D%5B%5D=10&work_search%5Bother_tag_names%5D=&work_search%5Bquery%5D=&work_search%5Blanguage_id%5D=&work_search%5Bcomplete%5D=0&tag_id=Alternate+Universe+-+Writing+*a*+Publishing), today's issue of [the Boston Globe](http://www.bostonglobe.com/), the interactive story ["Harmonia" by Liza Daly](http://ifarchive.org/if-archive/games/competition2017/Harmonia/index.html),  and this [PDF article in a medical journal](http://www.cpn.or.kr/journal/view.html?volume=16&number=2&spage=197). They have different needs and different modes of representation.

We need to make sure that all of our choices work in those cases as well. And that complicated the simplicity of "conceiving of "web publications"... as extant and enhance-able".

(For the sake of argument, I'm not giving edge  cases as examples, but only URLs where we'd all agree, and the authors would as well, that the resources in question are publications on the web.)

**Comment by HadrienGardeur** [comment]
> Is only true if our focus is entirely user agent focused. I worry about production.

@deborahgu sorry but I have to disagree on that one. EPUB is infinitely worse than a simple JSON manifest, yet millions of publications have been created in that format.

The Web in 2018 is also very far from the early days, where everything was indeed represented on a single HTML page. Take a look at any websites these days and you'll notice dozens of requests for JS, CSS, images, videos...
I don't think that a simple JSON manifest adds a lot to that complexity or goes against the "model of the Web".

**Comment by BigBlueHat** [comment]
> we all agree that web publications exist today, but I don't think we can pick a set of publications that do things in a certain way and say those show all needs can be easily met.

Apologies. It was not my intent to overly narrow the field by using those three publications as examples. There are some clear overlaps in needs/experiences of those three which makes finding possible technical improvements easier than a more "zoomed out" vantage point might. However, I certainly don't intend to narrow the scope of "publications" to book-like-things--especially given that I work for a scientific publisher where single-primary-resource publications are core to what we do.

The examples you gave are great, and worth considering in light of any approach we take.

**Comment by deborahgu** [comment]
@HadrienGardeur I'm afraid you're disagreeing with something I never said! As I've said in this thread, I honestly don't care whether we decide on JSON, HTML -- or for that matter, YAML or Markdown or m4 -- as long as it's easy to author, which to mean, includes having all descriptive metadata in one place.

I solicit you to post an example with a simple JSON manifest as @BigBlueHat did above, so that all members of the WG, including those who don't have enough knowledge about the fine technical details under discussion, can compare the various proposals on offer and decide if any of them would be overwhelmingly difficult (or quite simple!) to author.

**Comment by HadrienGardeur** [comment]
I also created static examples that could be useful.

Here's Moby Dick:

- Entry page available at: https://hadriengardeur.github.io/webpub-manifest/examples/MobyDick/
- WAM style manifest: https://hadriengardeur.github.io/webpub-manifest/examples/MobyDick/manifest.webmanifest
- RWPM style manifest: https://hadriengardeur.github.io/webpub-manifest/examples/MobyDick/manifest.json

And there's also "Why's Poignant Guide to Ruby":
- Entry page available at: https://hadriengardeur.github.io/webpub-manifest/examples/why/
- WAM style manifest: https://hadriengardeur.github.io/webpub-manifest/examples/why/manifest.webmanifest
- RWPM style manifest: https://hadriengardeur.github.io/webpub-manifest/examples/why/manifest.json

**Comment by BigBlueHat** [comment]
@HadrienGardeur are the entry page URLs the same as the publication address (URLs)? The publication address is absent from most of the manifest examples above.

In JSON-LD, it would (presumably) be expressed as the top-level `@id` value.

In the case of the WAM, there is no equivalent expression. The [`start_url` member](https://w3c.github.io/manifest/#x8-10-start_url-member) is the "URL that the developer would prefer the user agent load when the user launches the web application" and "is purely advisory, and a user agent MAY ignore it or provide the end-user the choice not to make use of it." Additionally, if the `start_url` is improperly expressed, the processing steps fallback to the document URL--which highlights the WAM augments an existing Web App, not the other way around.

This much older [Moby-Dick](https://dauwhe.github.io/html-first/Moby-Dick/) example built by @dauwhe and I explores [using a WAM](view-source:https://dauwhe.github.io/html-first/Moby-Dick/manifest.json) (and the mandatory ServiceWorker) to (if the user wants) install the Web (publication) App to a home screen. The ["binding" is expressed in HTML](view-source:https://dauwhe.github.io/html-first/Moby-Dick/) and processed via the [ServiceWorker](view-source:https://dauwhe.github.io/html-first/sw.js).

The "installed" experience is best on desktop--which can be done from the dev console (I'll try and find instructions and/or screenshots for that soon).

**Comment by HadrienGardeur** [comment]
@BigBlueHat I've created these examples a while back, but for RWPM the `identifier` element is tied to `@id` through the default context document. This could be a good fit for the entry page. 
In RWPM, identifier isn't necessarily meant for this (it's more in line with the canonical identifier from our WP infoset) but the entry page URL works fine as an identifier.

For the WAM, although there's no exact equivalent, the `start_url` is still conceptually the closest thing to an entry page.
For these examples, if you use the link for the viewer on the entry page, you get a Service Worker as well. It automatically handles caching for the default reading order and the list of resources as well.

I also have an older example that ties all of these things together (installable as a Web App with all the proper icons, WAM fields and WP to trigger the install dialog on Chrome mobile), I'll dig this.

**Comment by mattgarrish** [comment]
> How are these [infoset extraction] steps avoided?

I wouldn't even call them infoset extraction steps. I was reacting to bringing in things like generating the accessibility tree and CSS cascades. Are these really major concerns in determining the serialization of the manifest? It feels like premature optimization that may not even come into play.

We have better criteria to evaluate complexity, like ability and complexity of expressing the information needed through native mechanisms v. JSON. Meta tags fall back into the epub trap of un-nestable metadata, and parsing RDFa/microdata is (arguably) more complex to write and read than a JSON data structure. (XML anyone? ;)

Discovering the manifest is only one step in the process. And if the script tag containing the manifest is the last tag before the closing body, what has happened up until that tag is discovered? (Asked rhetorically.)

**Comment by deborahgu** [comment]
Thanks, @HadrienGardeur and @BigBlueHat!

@TzviyaSiegman and @GarthConboy, one thing that I think would bring in really valuable feedback (and engagement) from the rest of the WG, would be to as people to compare 

* [Benjamin's and Dave's example](https://github.com/WileyLabs/no-can-transclude/blob/master/moby-dick-from-epub-samples/package.opf), which is a standard opf file
* [Hadrien's WAM manifest](https://hadriengardeur.github.io/webpub-manifest/examples/MobyDick/manifest.webmanifest)
* [Hadrien's RWPM manifest](https://hadriengardeur.github.io/webpub-manifest/examples/MobyDick/manifest.json)
* @iherman, your idea of encoding the json in a script element on the landing page isn't represented here. 
* Any other options on the table? 

and be willing to discuss (in the meeting or in email) if any of them have any problems or features from an authoring perspective. I'd especially love to hear from WG members who work on the non-programmer side of ebook production, especially at the smaller organizations with fewer technical resources.

I acknowledge that this won't solve the technical issues under dispute, and there are other concerns from an authoring perspective.  (I see big cons on both of the major sides: Benjamin's example relies on polyfills, which IMO *always* cause accessibility failures and developer headaches in the wild; Hadrien's example relies on every page having navigation hard coded into the HTML.) Neither choice is perfect, and there are valid concerns from all comers. But it would be nice to resolve the authoring question @RachelComerford and I have raised.  If we're being overly fearful, that's a win!

**Comment by HadrienGardeur** [comment]
> Hadrien's example relies on every page having navigation hard coded into the HTML.

@deborahgu that's not entirely true.

If you're using a WP enabled UA (such as the basic viewer that I link to), there's no need to have hard coded navigation in HTML.
Also: this is not really tied to any serialization or way we embed our infoset. We'll face that same issue with any option that we discuss.

Here's another example where navigation is dynamically injected by a JS script, based on the manifest:  https://hadriengardeur.github.io/webpub-manifest/examples/progressive-enhancements/

(This example is also a valid Web App and has a Service Worker to handle offline reading).

**Comment by HadrienGardeur** [comment]
Thanks @iherman for this mashup of my previous example.

That said, I think we need to make a distinction between the TOC and the default reading order, which is not the case in your example.
The default reading order is a requirement, while the TOC is optional.

In a more complex use case where these two are different, we should have unique roles to identify them.
In a situation where the default reading order and the TOC are the same thing, we should still IMO include two roles and not just `doc-toc`, in order to avoid confusion.

It's also worth pointing out that JSON as a separate resource vs JSON embedded in HTML are not necessarily two exclusive options.
The JSON could still contain a canonical link to a resource, even when it's embedded in HTML.

**Comment by llemeurfr** [comment]
@RachelComerford 

> it is easier and cheaper for me to find HTML coding resources than JSON coding resources

Looking at the JSON sample @iherman crafted; let's imagine that a JSON template expressing such structure is made available to people who just have to learn how to add information about a publication here: do you think that it would be a difficult / expensive task? more difficult / expensive than crafting metadata in HTML like what we find in https://idpf.github.io/epub-guides/schema-org-integration/#h.2urrswf91171 ?

**Comment by BigBlueHat** [comment]
> Benjamin's and Dave's example, which is a standard opf file

@deborahgu the OPF file isn't used at all in our example. All the interaction/understanding is built from the `toc.xhtml` nav document (which was renamed to `index.html` to make it the entry page). The OPF file is "left overs" because those files used to be an EPUB. 😄 It's not used at all by our examples.

**Comment by BigBlueHat** [comment]
> Benjamin's example relies on polyfills

It only relies on a polyfill type script to provide an overlaid reader. If you were to remove that script tag, the publication is still completely browse-able, readable, etc...it just looses the next/prev progression experience provided by the polyfill.

Any polyfill should be disposable in the long term and any content it leans on should be written to be accessible and the UI (if any) which it adds should also be presented accessibly. These are things I care about which may not always come through in the demos--for which I apologize.

**Comment by BigBlueHat** [comment]
> I presume the idea is to use the <meta> elements in the header instead.

The *metadata* could be expressed via `<meta>`, RDFa, JSON-LD (inline or out-of-band), Microdata, or anything else we determine as a group is the best place to express, find, and process the metadata.

The core distinction of the HTML-first examples is that it puts the "binding" components of the infoset (primary resources and dependencies) into a space where the browser already understands how to "bind" things (i.e. <script>, <a>, <link>, <style>, <iframe>, etc) vs. redefining all those things in a descriptive-only JSON (or any other non-hypermedia) format.

Perhaps that's the clarification that's been getting lost in the weeds?

**Comment by iherman** [comment]
@BigBlueHat can you provide the equivalent of the two other solutions of https://github.com/w3c/wpub/issues/193#issuecomment-388283248? I must admit I do not understand what you say:-(

(I am happy to put up that example on the same github area, if it is better)

Thanks.

**Comment by BigBlueHat** [comment]
@iherman the key difference (which I'm happy to craft and post somewhere) would be that the `resources` member would be expressed in HTML as dependencies (via `<script>`, `<style>` etc) of the resources that depend on them. It is possible to aggregate these dependencies for offline use via ServiceWorkers as @dauwhe and I showed originally in https://dauwhe.github.io/html-first/Moby-Dick/ (which currently does the caching automatically on SW registration vs. via a user-initiated request).

So, these lines would be removed:
https://github.com/iherman/misc-notes/blob/master/wp-manifest-experiments/manifest_script/mobydick.html#L22-L28

**Comment by iherman** [comment]
I think it would be more helpful for everyone if you created a separate file, not only with the removal of the data from the script, but also what and how you would replace it... 

None of the examples are final, obviously (see also [Hadrien's comments](https://github.com/w3c/wpub/issues/193#issuecomment-388286056)), but that is besides the point at this moment. And let us not have any javascript or whatever there; I guess, mainly for @deborahgu and @RachelComerford the important point is _what_ should be produced by whoever produces the publication...

Thx.

**Comment by BigBlueHat** [comment]
> I've updated both RWPM examples to point to the entry page in identifier.

@HadrienGardeur thanks. Is the `identifier` in `metadata` what you changed? There still doesn't seem to be a top-level `identifier` (i.e `@id`) for the JSON-LD. The consequence of that is that these statements become detached when actually processed because they'll point to a blank node (i.e. auto-generated, local-to-that-request/moment identifier) rather than to the "publication address."

I'm also operating on the fact that "publication address" equates to the URL identification of the publication and not merely where one gets an "entry point/page."

If "publication address" does *not* provide the URL identification of the publication, then perhaps that's at the core of some of our misunderstandings.

**Comment by deborahgu** [comment]
So here's a question:

From the point of view of the  publisher's production staff, there's not much difference between

```
    "title": "Moby-Dick",
    "author": "Herman Melville",
    "language": "en",
    "modified": "2018-02-10T17:00:00Z"
```

and (copied directly from my library network's catalog)

```
    '"Publication Info": "New York : Pearson Longman, c2007.",
    '"Physical extent": "xxviii, 660 p. : ill., map ; 23 cm.",
    '"ISBN": "0321228006 (alk. paper)",
    '"ISBN": "9780321228000 (alk. paper)",
    '"ISBN": "0205514081 (pbk.)",
    '"ISBN": "9780205514083 (pbk.)",
    '"Series": "A Longman critical edition Longman critical edition.",
    '"Bibliography": "Includes bibliographical references.",
    '"Subject": "Melville, Herman, 1819-1891. ",
    '"Subject": "Moby Dick. Ahab, Captain (Fictitious character) -- Fiction. ",
    '"Subject": "Whaling ships -- Fiction. ",
    '"Subject": "Ship captains -- Fiction. ",
    '"Subject": "Whaling -- Fiction. ",
    '"Subject": "Whales -- Fiction. ",
    '"Subject": "Psychological fiction. ",
    '"Subject": "Sea stories.",
    '"Added Author": "Bryant, John, 1949- ",
    '"Added Author": "Springer, Haskell S.",
    '"OCLC/Vendor #": "69671881",

```

Is there anything about any of these solutions which would *prevent* them from also being used to storefor the purely descriptive metadata which isn't that a necessary component of the infoset?

**Comment by HadrienGardeur** [comment]
> @HadrienGardeur thanks. Is the identifier in metadata what you changed? There still doesn't seem to be a top-level identifier (i.e @id) for the JSON-LD. The consequence of that is that these statements become detached when actually processed because they'll point to a blank node (i.e. auto-generated, local-to-that-request/moment identifier) rather than to the "publication address."

I don't want to rehash another discussion again, we've already went through this on the mailing list several months ago.

I would advise that you look at the example that @iherman has tweaked instead, since this directly solves your issue with how `@id` is expressed: https://github.com/iherman/misc-notes/blob/master/wp-manifest-experiments/separate_manifest/mobydick.json

> I'm also operating on the fact that "publication address" equates to the URL identification of the publication and not merely where one gets an "entry point/page."
>
> If "publication address" does not provide the URL identification of the publication, then perhaps that's at the core of some of our misunderstandings.

In RWPM we only have an identifier and there's no such thing as an "entry page". `identifier` is meant to contain a URI that uniquely identifies the publication.

That said, given the extensibility model of RWPM, there are many options in addition to using `identifier`, including but not limited to:

- using terms from schema.org
- using a link with the right rel value to identify an entry page

IMO, this discussion is out of scope and we should focus instead on the benefits (if any) of embedding JSON-LD in a script element vs linking.

On my side, the list looks like this:
- Pro: Keeping things simple for single page publications
- Pro: Can be indexed by search engines if we align with that they support
- Con: harder to access (no direct URL)
- Con: harder to reference (must rely on a fragment) 
- Con: harder to parse (need to parse HTML first)

I don't think that embedding the infoset items reduces redundancy significantly aside from the table of contents. 
The default reading order can be very different from a TOC and the list of resources (as we define it in our infoset) doesn't have any equivalent in HTML. They're also "hidden" list and we've had pretty bad experience in EPUB with attempts to hide such lists in the navigation document.

**Comment by BigBlueHat** [comment]
>I don't want to rehash another discussion again, we've already went through this on the mailing list several months ago.
> ...your issue with how @id is expressed

Sorry, that's not what I was referring to. Whatever you map `@id` to, there MUST be one at the top level or the subject of all those statements is "unknown" (i.e. a blank node). @iherman's fixes that by adding the publication address as the subject/identifier of [his example document](https://github.com/iherman/misc-notes/blob/master/wp-manifest-experiments/separate_manifest/mobydick.json).

> The default reading order can be very different from a TOC

Would love to see some examples of these if anyone has one. The HTML-first approach could be modeled such that the TOC is not the "reading order" expression, but rather references points in that list (and displays/loads them) as can be seen in this example (which uses *no* JavaScript): https://wileylabs.github.io/no-can-transclude/carol/

The reading order is defined (in this case) by a series of `<iframe>`'s and referenced from a TOC via fragment identifiers. Obviously that (like everything else everywhere) has it's own set of limitations, but it does get closer to presenting the sort of HTML-based modeling of a "binding document" on which we could build interesting affordances (via just a handful of new things...hopefully).

> and the list of resources (as we define it in our infoset) doesn't have any equivalent in HTML

It does, but that "list of resources" (which I've been calling dependencies here because the relationship seems clearer) are referenced "in place" from the content document(s) which use them.

What that practically does is put the responsibility of "gathering" that list upon the browser/reading-systems (via an optional--depending on the use case--and possibly asynchronous) "gathering" process rather than on the author/publisher who would have to keep that list in sync by gathering all those things and listing them in the JSON prior to publication.

**Comment by iherman** [comment]
@deborahgu, I hope I understand your question when I try to answer...

Is your question whether it is possibly to add the additional metadata items to the JSON encoding of the infoset (whether it is in a separate file or embedded in the HTML file)?

The answer is... 'yes' and 'not without some further steps' (I know, this is the most irritating answer possible:-).

'Yes', in the sense that the JSON syntax is extremely flexible, so it can include all those terms. All JSON parsers would accept it without problems.

'No', in the sense that: are these terms to be standardized by this Working Group? That would open up a giant Pandora's Box, in view of the large variety of metadata terms that are used out there, and also their specificities to particular areas of publishing. I do not think we should get into such standardization; there are other bodies, like EDItEUR, that do that. But if we do not do such term standardization, the question is whether it is a good idea to allow any kind of metadata added to the JSON data without further control; that may raise issues of checking of a WP, for example.

However, at the moment, the infoset spec says that it should be possible to [_extend_](https://w3c.github.io/wpub/#wp-extensibility) the infoset which, translated to JSON (or HTML, for that matter) means that such extension should be possible either through some control over the vocabulary or linking to external, but separate, metadata files. The technical details of this is still to be worked out, though.

(If we work within the additional framework of JSON-LD, for example, that may means that and additional `@context` should be provided specifying, eg, the ONIX terms; if this is added to the manifest then those extensions can be used without further ado. But that would require to have an RDF reference to ONIX terms, which may not really exist... It is an example where, probably, linking to an external ONIX file may be better.)

**Comment by llemeurfr** [comment]
In reply to @deborahgu,regarding the the possibility of a direct copy from a library catalog to JSON or HTML ("any of these solutions"). 

As you can see, the data structure of JSON is very similar to the structure of your library catalog : this is raw structured information, and translation is immediate. Moving such library structure to HTML is less immediate (use of non-recursive meta elements or use of content elements for embedding a metadata structure using on of the discussed solution, RDFa, microdata ...). 

About extensibility: JSON is extensible, as @iherman states, but even more interesting, JSON validation tools, i.e. JSON Schemas (a specification + well know software tools), also allow for that type of free extension. A schema can check all required metadata and their value, check that optional metadata have a proper value, and let free external metadata be added without choking. On the HTML side, people can freely extend a metadata vocabulary, but there is no validation mechanism for required and optional metadata: somebody has to code something specific to enforce such rules (a full epubcheck).  

It's not to say that ANY metadata should be embedded in JSON; only JSON serialized metadata can. XML serialized metadata must be expressed somewhere else and can be referenced from the JSON manifest (the nervous system of the publication). 

I think that the decision to externalize every metadata that is not in the infoset, a decision that was taken in the early years of EPUB, is to be revised. But this should be discussed in another issue. 

I also believe that the initial question of this very long thread has been answered: splitting the infoset in JSON + HTML (meta tags or content) is boring for both authors (information to put in two places with different serializations) and developers of user agents. On the other side, embedding the JSON manifest containing the full infoset in an HTML header is possible and is even interesting for publications with a single resource.

**Comment by HadrienGardeur** [comment]
> Sorry, that's not what I was referring to. Whatever you map @id to, there MUST be one at the top level or the subject of all those statements is "unknown" (i.e. a blank node). @iherman's fixes that by adding the publication address as the subject/identifier of his example document.

@BigBlueHat that's exactly what we've discussed on the mailing list before and @iherman even hinted at a solution to this issue based on JSON-LD 1.1.

> The reading order is defined (in this case) by a series of <iframe>'s and referenced from a TOC via fragment identifiers. Obviously that (like everything else everywhere) has it's own set of limitations, but it does get closer to presenting the sort of HTML-based modeling of a "binding document" on which we could build interesting affordances (via just a handful of new things...hopefully).

Yikes, I'll have nightmares with that paragraph.

The reading order is the most important item in our infoset IMO and a list of `<iframe>` elements referenced by a a TOC using fragment sounds like the kind of examples that would scare any reading system developer away.
I also have a very very hard time believing that anyone would rather author something that complex instead of providing a straightforward list in JSON.

(Any thoughts on this @JayPanoz?)

>  It does, but that "list of resources" (which I've been calling dependencies here because the relationship seems clearer) are referenced "in place" from the content document(s) which use them.
>
> What that practically does is put the responsibility of "gathering" that list upon the browser/reading-systems (via an optional--depending on the use case--and possibly asynchronous) "gathering" process rather than on the author/publisher who would have to keep that list in sync by gathering all those things and listing them in the JSON prior to publication.

OK, let me get this straight, with your suggestion the list of resources:

- would potentially be scattered across resources
- would require the UA to fetch every resource in the reading order (which is expressed using URIs to fragments + `<iframe>` elements)
- would not be explicitly listed in each resource (links), but would rely instead on the ability for the UA to intercept network requests

Is that accurate?

**Comment by JayPanoz** [comment]
Well, all I can say at this point is that I keep seeing the “ease of authoring” concept/argument used in (too) many ways that are partly or completely disconnected from what I have witnessed in the EPUB trenches for 6 years, in a lot of discussions. 

If you really want to take graphic designers and independent publishers into account, then it means “if the authoring tools they’re using don’t implement it, they won’t use it.”

A few examples:

- they will use Calibre and/or Sigil to add metadata – that’s XML;
- it’s already quite a feat to make them use the correct output markup in InDesign – which is why so many files have everything as a `<p>` or a `<span>` – that’s HTML;
- they will use Sigil to generate the EPUB3 nav even if InDesign can manage that – that’s HTML;
- if a feature is not implemented, they simply don’t use it, albeit it would be 1 change in the HTML documents and 2 extra lines of CSS;
- for interactive books, they will simply import an existing lib/plugin – that’s JS;
- the most popular question during training sessions, by very far… “Is there an app for this?” – that’s their artisanal workflows.

I’m stopping there but there is so much more. The sad truth is HTML is already asking way too much to a significant number of the “spec users” anyway, and they will rely solely on authoring tools – which tend to be underrepresented in a lot of discussions related to authoring, unfortunately.

Note that even if they master all of HTML, CSS and JS, a significant portion of authors are very likely to use tools that ease their lives, cf. [PWABuilder](https://www.pwabuilder.com) by Microsoft.

**Comment by RachelComerford** [comment]
Forgive me but I am going to ask that we go back to basics here for a minute... what is the problem that these 2 proposed solutions:

1. Is it acceptable to use HTML for the serialization of some infoset items, 
2. or should it all be in separate (JSON) file? 

are meant to solve?
Can someone give a rundown of the alternatives that we're considering?

**Comment by iherman** [comment]
@RachelComerford, you are right, it is a good idea to pause and maybe formulate the high level choices...

I think a way to formulate the two positions, by formulating them to the extreme might be as follows (and this is obviously _my_ view of things at 6am in the morning...)

1. _All_ infoset items should be defined in JSON (possibly JSON-LD). It is a bit like what we have today in EPUB; all various information should be in one place with a unified syntax, including not only metadata, but the list of resources, ToC, whatever. That unified syntax should be in JSON because that is the syntax used by Web Developers, because it is very easy to "parse" it and work with the result, as opposed to an XML syntax which is way more complex to handle. The infoset file is accessed via a special `<link>` in what is often called an “entry page”, which is in HTML.
2. _All_ infoset items should be expressed via a combination of HTML elements (not only `<meta>` or `<link>`, but also, say `<nav>`) and possibly some information conveyed via HTTP. All (or most of?) these informations are collected into that entry page.

There are some intermediate proposals floating around which may make this less clear-cut. No. 1 above, for example, may lead to redundancy of information (which is also present in EPUB) mainly in terms of the list of resources that make a specific WP, or the ToC: there are some entries in the current draft that propagates the re-use of an HTML ToC to extract the list of resources to mitigate that. 

The extra proposal that I have put forward is somewhere between the two (although I admit closer to No.1), namely to use the JSON syntax as described in No 1, but incorporate it into the header of the entry page as using the `<script>` tag. That would open the way of some sort of a possible compromise whereby _some_ infoset items _could_ be expressed via good-old HTML elements within the entry page, but other entries could go to JSON if it is more natural or semantically clearer. But even that compromise solution leads to the same dichotomy: should the `<script>` element be the _**only**_ way to use JSON or would it be an optional usage of No.1?

What is more difficult is to get a clear set of pros and cons. I will try to get _some_ below, knowing that I will not make full justice to any of the two sides...

* A JSON formulation is (much) simpler. We do have the experience (good and bad...) with EPUB, and the infoset is defined, in the abstract sense to a, say, 75-80% level. We can have discussions on syntactic details, whether to use JSON-LD or not and, if yes, how, but none of these are really major issues. Properly defining the HTML approach seems to be much more costly in time and energy: we (I?) do not have a clear idea on what exactly should be done and how.
* The JSON approach may lead to a redundancy that was also present in EPUB, eg, for the list of resorces. Whether that redundancy is really really bad, or we can live with it is not clear and opinions clearly diverge on this. (There is a big difference, in this sense, whether the WP consists of a lot of HTML content resources or only one.)
* It should be possible to define the entry page with all (most of?) the items in HTML in such a way that a browser that does not know anything of WP-s could make some sense of it. If that is where the "spine"/ToC content is, if some of the traditional HTML-syntaxed metadata are (on dates, licenses, titles, etc.) than this may be doable and that makes sense. (That being said, it is much less clear whether that statement is true even in cases when there are lots of resources in a WP.) A JSON syntax has no meaning for a browser that does not know WP-s.
* There are some tricky issues around, e.g., security (what scripts/files can I, should I, may I load, for example) that have an immediate answer if only HTML is considered (because these are handled by existing standards) whereas it is not entirely clear what happens to that if all extra information are in a separate JSON structure.
* The HTML only approach stretch the limits of the current HTML specification, which was _never_ defined with the concept of several documents being in the same logical place. A proper way of doing some of the things we need may be to try to slightly modify the HTML specification itself, which is an extremely tall order (which I consider as unattainable).

I am sure I will get lots of flames now... Fire off!

**Comment by BigBlueHat** [comment]
> the kind of examples that would scare any reading system developer away.

It's been my operating understanding that the primary audience for this specification is for Web browsers. I've also tried to based my proposals on "web publications" (lowercase on purpose) as they exist on the Web today. My goal has been to find ways to enhance *existing* web publications by adding the least amount of tech required to add key affordances (https://github.com/w3c/wpub/labels/topic%3Aaffordances) currently unavailable to the human reading the publication.

Perhaps its just that our objectives differ?

**Comment by BigBlueHat** [comment]
@iherman your summaries in https://github.com/w3c/wpub/issues/193#issuecomment-388696570 are spot on. The [`<script>`-based data block](https://www.w3.org/TR/html5/semantics-scripting.html#data-block) for metadata is completely fine for encoding HTML-based *metadata* about the publication.

The core sticking point seems to be around the expression and processing of the "binding"--i.e. the thing that defines the multi-resource/document experience.

I'd like to shorten the distance between publication address and readable/experience-able publication. Since the Web (browser) does HTML by default, that (to me) means building up from that foundation.

**Comment by RachelComerford** [comment]
This is what I understand the problem and a potential compromise/solution to be. Interested in hearing feedback:

**Problem**: There is information that must be available via the infoset and this needs to be coded and housed somewhere

**Proposed Solution**: Some information (what exactly is TBD) will be housed within the HTML, some within JSON. That JSON may or may not be a separate file.

**Alternatives Considered**:
1. All infoset items should be defined in JSON (possibly JSON-LD). It is a bit like what we have today in EPUB; all various information should be in one place with a unified syntax, including not only metadata, but the list of resources, ToC, whatever. That unified syntax should be in JSON because that is the syntax used by Web Developers, because it is very easy to "parse" it and work with the result, as opposed to an XML syntax which is way more complex to handle. The infoset file is accessed via a special <link> in what is often called an “entry page”, which is in HTML.

2. All infoset items should be expressed via a combination of HTML elements (not only <meta> or <link>, but also, say <nav>) and possibly some information conveyed via HTTP. All (or most of?) these informations are collected into that entry page.

**Comment by llemeurfr** [comment]
@RachelComerford maybe this issue could be closed if we can extract from it the TOC issue.

There were  looong discussions last year about the representation of the reading order (ex-spine), the representation of a human-friendly Table of Contents and the need for machine readable navigation. https://github.com/w3c/wpub/issues/9 is an example of such loong threads. @HadrienGardeur 's comment also makes reference to the TOC issue. And this issue is far from being solved.

So I suggest to keep the TOC out of scope of this issue and rephrase:

**Proposed Solution**: The TOC being kept out of scope for now, the infoset is housed within a JSON manifest. That JSON may be embedded in the entry page if the publication has a single resource.

**Comment by BigBlueHat** [comment]
Good summary, @RachelComerford.

@llemeurfr this thread covers far more than the role of the "TOC issue" and has (afaict) helped clarify that "infoset" currently encompasses inert, descriptive metadata values (i.e. [descriptive properties](https://w3c.github.io/wpub/#infoset-meta)) as well as active, hypermedia-style "binding" expressions (i.e. [structural properties](https://w3c.github.io/wpub/#infoset-structure)). 

Where one wants to see those things collected seems to pivot on the expected processing models of either a Web browser (+/- future publication affordances) or a more EPUB-style Reading System.

@RachelComerford your proposed solution summary is a good one:
> Proposed Solution: Some information (what exactly is TBD) will be housed within the HTML, some within JSON. That JSON may or may not be a separate file.

Roughing up an example now.

**Comment by BigBlueHat** [comment]
I've just sent a PR with some examples (no spec changes), one of which is a minimal, `<nav>` fallback (as currently drafted) HTML entry page: https://github.com/w3c/wpub/pull/196/commits/15682e5ffae454ae963f5bbe019bbcd1460d1ff1

**Comment by iherman** [comment]
@BigBlueHat good idea to have moved the example to the wpub repo (I have just merged it).

Looking at https://github.com/w3c/wpub/blob/master/experiments/html-schema-org-json-ld/index.html, and comparing it with the other example, I do not see how you intend to represent the list of (other) resources, that appear in the other manifest.

**Comment by HadrienGardeur** [comment]
> It's been my operating understanding that the primary audience for this specification is for Web browsers. I've also tried to based my proposals on "web publications" (lowercase on purpose) as they exist on the Web today. My goal has been to find ways to enhance existing web publications by adding the least amount of tech required to add key affordances (https://github.com/w3c/wpub/labels/topic%3Aaffordances) currently unavailable to the human reading the publication.
>
> Perhaps its just that our objectives differ?

I've also been building on top of existing Web Publications, and the example for "Why's poignant guide to Ruby" is a perfect illustration of that since I only added an entry page and a manifest instead of re-hosting and modifying the content:
- Entry page available at: https://hadriengardeur.github.io/webpub-manifest/examples/why/
- WAM style manifest: https://hadriengardeur.github.io/webpub-manifest/examples/why/manifest.webmanifest
- RWPM style manifest: https://hadriengardeur.github.io/webpub-manifest/examples/why/manifest.json

For the primary audience, my take on this issue is that we should design something that works well for every type of UA, not just browsers.

That said, I don't think that your argument really holds up @BigBlueHat. `iframe` elements being referenced through fragment identifiers sounds like the worst part of EPUB (the whole id/idref mess in metadata and manifest/spine) and isn't easier to handle for a browser than a simple list in JSON.

As @iherman has pointed out, your example lacks a list of resources and some of your [previous comments](https://github.com/w3c/wpub/issues/193#issuecomment-388467714) about this infoset item were IMO hard to understand.

I already posted a summary in [a previous comment](https://github.com/w3c/wpub/issues/193#issuecomment-388575454), could you confirm that I understood things correctly?

> OK, let me get this straight, with your suggestion the list of resources:
>
> - would potentially be scattered across resources
> - would require the UA to fetch every resource in the reading order (which is expressed using URIs to fragments + <iframe> elements)
> - would not be explicitly listed in each resource (links), but would rely instead on the ability for the UA to intercept network requests

**Comment by HadrienGardeur** [comment]
> I am not sure that "restricting" the JSON expression to a script element is only relevant for a single-HTML case. I actually wonder, in view of the fact that this is the only way the JSON-LD content is digested by schema.org, whether this form should not be the preferred one.

@iherman nothing is digested by schema.org, it's only the place where things are defined.

But JSON-LD with schema.org terms is indeed digested by a number of search engines. It's what Google recommends and Bing recently added support for it as well.

These search engines can only process JSON-LD contained in a `<script>` element for now, but it doesn't have to be static. If a script dynamically injects JSON-LD in a page, this will be processed as well.

For our use case, this might be the optimal outcome:

- if you include JSON-LD statically on the entry page, only the entry page will contain the markup
- if you rely on JS to check the presence of `<link rel="publication">` and use this to dynamically inject JSON-LD in a `<script>` element, all resources of the publication could potentially contain this markup, without any risk regarding duplication/inconsistencies

The second option is often used for SEO optimizations, as it provides an easy way to serve static content with dynamic metadata.

As a side-note, once again I really think that we're having two separate discussions on this issue and it's making things more confusing than they should be:
- which infoset items MAY/SHOULD/MUST be expressed in JSON vs HTML is one issue
- whether we should embed JSON on the entry page or link to the manifest as a separate resource is a different issue

**Comment by BigBlueHat** [comment]
> @BigBlueHat good idea to have moved the example to the wpub repo (I have just merged it).

Oh! That'll work. I'd been thinking we'd use the PR review system to discuss them, but I guess in the repo we can still reference line numbers, so that'll work.

I do think it'll help to see actual experiments as code vs. prose.
 
> Looking at https://github.com/w3c/wpub/blob/master/experiments/html-schema-org-json-ld/index.html, and comparing it with the other example, I do not see how you intend to represent the list of (other) resources, that appear in the other manifest.

Dependencies are gathered from the primary resources that depend on them. Just as when you GET an HTML file, the browser will subsequently GET all the JS and CSS (etc) that it references.

This does not (by design) include a more package-focused, publication-wide list of all the things that ever thing in the publication depends on. The Web is built from just-in-time referencing and retrieval. This follows that design pattern.

**Comment by BigBlueHat** [comment]
> @BigBlueHat about #193 (comment), we must now stop with generalities and prototypes and close this issue (65 comments) with a clear answer.

This Working Group's been using GitHub for discussion as well as issues, so while I agree that this one has gotten rather long (and am working to break out specific, actionable issues), I don't feel it's closable while there's still things to extract and/or discuss further.

> The spec states clearly in https://w3c.github.io/wpub/#infoset that the infoset covers both descriptive and structural properties, nothing new there.

The "infoset" term does currently encompass both types of properties and #197 has been created to explore those two different groups in light of the discussions here.

> The only part of the infoset (as currently defined) on dispute is the ToC. In your latest experiment, all other properties are in the JSON structure. So you are proving my point, thank you.

The dispute seems more about the use of the ToC to potentially provide the reading order and primary resource list. Additionally, there are open questions around how dependencies of primary resources are expressed:
 - either explicitly--as in a [resource list](https://w3c.github.io/wpub/#wp-resource-list) of all the things
 - or implicitly--as dependencies gathered/requested as needed by the resource that depends on them.

There is much more to discuss and explore, but it doesn't need to stay on this issue, and I concur that narrower topics should be made wherever possible.

More specific issues to follow.

**Comment by GarthConboy** [comment]
Resolved on call: "Proposed resolution: The infoset mostly resides within a JSON manifest (WP manifest). That JSON may optionally be embedded in the entry page rather than a standalone file referenced by <link> from the entry page.  It may be supported to allow some infoset items to reside in HTML of the entry page, if information duplication issues can be sufficiently avoided."
